ckan/lib/dictization/__init__.py

```diff
@@ -132,7 +131,11 @@ def table_dict_save(table_dict, ModelClass, context):
             params = dict((key, table_dict.get(key)) for key in constraint)
             obj = session.query(ModelClass).filter_by(**params).first()
             if obj:
-                break
+                if 'name' in params and getattr(obj, 'state', None) == State.DELETED:
+                    obj.name = obj.id
+                    obj = None
+                else:
+                    break
```

Comment:
IIUC this will clear a datasets name when it's deleted, but if a dataset was deleted before this change is made the change will fail when postgres enforces the unique name constraint.

Instead of this I suggest catching the postgres failure and rename the deleted dataset before retrying the one we're trying to save.